### PR TITLE
 initial support for boringssl build under AOSP

### PR DESCRIPTION
**Pull request by chenbd**
There are some TODOs left and not sure how to deal with them.
Feel free to edit the commit when anything needs to be fixed.

### CODE REVIEW

File: src/openssl/app.c [hunk 1]
```diff
@@ -40,6 +40,10 @@
 #include <xmlsec/openssl/evp.h>
 #include <xmlsec/openssl/x509.h>
 
+#ifdef OPENSSL_IS_BORINGSSL
+#include "openssl_compat.h"
```
Comment (lsh123) [inline src/openssl/app.c]: No need for ifdef, just include it all the time

File: src/openssl/app.c [hunk 2]
```diff
@@ -71,7 +75,9 @@ xmlSecOpenSSLAppInit(const char* config) {
 #if !defined(XMLSEC_OPENSSL_API_110)
 
     ERR_load_crypto_strings();
+#ifndef OPENSSL_IS_BORINGSSL
```
Comment (lsh123) [inline src/openssl/app.c]: Please define it in openssl_compat.h as a noop

File: src/openssl/app.c [hunk 3]
```diff
@@ -126,8 +132,10 @@ xmlSecOpenSSLAppShutdown(void) {
     RAND_cleanup();
     EVP_cleanup();
 
+#ifndef OPENSSL_IS_BORINGSSL
```
Comment (lsh123) [inline src/openssl/app.c]: Same

File: src/openssl/evp.c [hunk 4]
```diff
@@ -315,13 +319,18 @@ xmlSecOpenSSLEvpKeyDup(EVP_PKEY* pKey) {
...
     ret = EVP_PKEY_up_ref(pKey);
     if(ret <= 0) {
         xmlSecOpenSSLError("EVP_PKEY_up_ref", NULL);
         return(NULL);
     }
 
     return(pKey);
+#else
+    (void)ret;
+    return EVP_PKEY_up_ref(pKey);
```
Comment (lsh123) [inline src/openssl/evp.c]: What if it fails?

File: src/openssl/kw_des.c [hunk 5]
```diff
@@ -34,6 +34,9 @@
 #include <xmlsec/openssl/crypto.h>
 
 #include "../kw_aes_des.h"
+#ifdef OPENSSL_IS_BORINGSSL
```
Comment (lsh123) [inline src/openssl/kw_des.c]: Same as above

File: src/openssl/openssl_compat.h [hunk 6]
```diff
@@ -18,11 +18,25 @@
  *****************************************************************************/
 #if !defined(XMLSEC_OPENSSL_API_110)
 
+#ifndef OPENSSL_IS_BORINGSSL
```
Comment (lsh123) [inline src/openssl/openssl_compat.h]: Could you please move all boringssl staff into its own section?

File: src/openssl/x509.c [hunk 7]
```diff
@@ -47,6 +47,9 @@
 #include <openssl/x509_vfy.h>
 #include <openssl/x509v3.h>
 #include <openssl/asn1.h>
+#ifdef OPENSSL_IS_BORINGSSL
```
Comment (lsh123) [inline src/openssl/x509.c]: Same as above

File: src/openssl/x509.c [hunk 8]
```diff
@@ -402,7 +405,11 @@ xmlSecOpenSSLKeyDataX509GetCert(xmlSecKeyDataPtr data, xmlSecSize pos) {
     ctx = xmlSecOpenSSLX509DataGetCtx(data);
     xmlSecAssert2(ctx != NULL, NULL);
     xmlSecAssert2(ctx->certsList != NULL, NULL);
+#ifdef OPENSSL_IS_BORINGSSL
```
Comment (lsh123) [inline src/openssl/x509.c]: Just cast to size_t

File: src/openssl/x509.c [hunk 9]
```diff
@@ -486,7 +493,11 @@ xmlSecOpenSSLKeyDataX509GetCrl(xmlSecKeyDataPtr data, xmlSecSize pos) {
     xmlSecAssert2(ctx != NULL, NULL);
 
     xmlSecAssert2(ctx->crlsList != NULL, NULL);
+#ifdef OPENSSL_IS_BORINGSSL
+    xmlSecAssert2(pos < sk_X509_CRL_num(ctx->crlsList), NULL);
```
Comment (lsh123) [inline src/openssl/x509.c]: Same as above

File: src/openssl/x509vfy.c [hunk 10]
```diff
@@ -185,7 +185,11 @@ xmlSecOpenSSLX509StoreVerify(xmlSecKeyDataStorePtr store, XMLSEC_STACK_OF_X509*
     X509 * err_cert = NULL;
     X509_STORE_CTX *xsc;
     int err = 0;
+#ifdef OPENSSL_IS_BORINGSSL
+    size_t i;
```
Comment (lsh123) [inline src/openssl/x509vfy.c]: Same as above

File: src/openssl/x509vfy.c [hunk 11]
```diff
@@ -719,7 +723,11 @@ xmlSecOpenSSLX509FindCert(STACK_OF(X509) *certs, xmlChar *subjectName,
                         xmlChar *issuerName, xmlChar *issuerSerial,
                         xmlChar *ski) {
     X509 *cert = NULL;
+#ifdef OPENSSL_IS_BORINGSSL
+    size_t i;
```
Comment (lsh123) [inline src/openssl/x509vfy.c]: Same as above

File: src/openssl/x509vfy.c [hunk 12]
```diff
@@ -836,7 +844,11 @@ xmlSecOpenSSLX509FindCert(STACK_OF(X509) *certs, xmlChar *subjectName,
 static X509*
 xmlSecOpenSSLX509FindNextChainCert(STACK_OF(X509) *chain, X509 *cert) {
     unsigned long certSubjHash;
+#ifdef OPENSSL_IS_BORINGSSL
+    size_t i;
```
Comment (lsh123) [inline src/openssl/x509vfy.c]: Same as above

File: src/openssl/x509vfy.c [hunk 13]
```diff
@@ -857,7 +869,11 @@ xmlSecOpenSSLX509VerifyCertAgainstCrls(STACK_OF(X509_CRL) *crls, X509* cert) {
     X509_NAME *issuer;
     X509_CRL *crl = NULL;
     X509_REVOKED *revoked;
+#ifdef OPENSSL_IS_BORINGSSL
+    size_t i, n;
```
Comment (lsh123) [inline src/openssl/x509vfy.c]: Same as above

File: src/openssl/evp.c (hunk 4 above)
Comment (chenbd) [reply src/openssl/evp.c]: guess it should not fail:
`
EVP_PKEY *EVP_PKEY_up_ref(EVP_PKEY *pkey) {
  CRYPTO_refcount_inc(&pkey->references);
  return pkey;
}
`

File: src/openssl/app.c (hunk 1 above)
Comment (chenbd) [reply src/openssl/app.c]: fixed

File: src/openssl/app.c (hunk 2 above)
Comment (chenbd) [reply src/openssl/app.c]: fixed

File: src/openssl/app.c (hunk 3 above)
Comment (chenbd) [reply src/openssl/app.c]: fixed

File: src/openssl/kw_des.c (hunk 5 above)
Comment (chenbd) [reply src/openssl/kw_des.c]: fixed

File: src/openssl/x509.c (hunk 9 above)
Comment (chenbd) [reply src/openssl/x509.c]: fixed

File: src/openssl/x509vfy.c (hunk 10 above)
Comment (chenbd) [reply src/openssl/x509vfy.c]: fixed to size_t

File: src/openssl/x509vfy.c (hunk 11 above)
Comment (chenbd) [reply src/openssl/x509vfy.c]: fixed to size_t

File: src/openssl/x509vfy.c (hunk 12 above)
Comment (chenbd) [reply src/openssl/x509vfy.c]: fixed to size_t

File: src/openssl/x509vfy.c (hunk 13 above)
Comment (chenbd) [reply src/openssl/x509vfy.c]: fixed to size_t

**Comment by lsh123** [comment]
I made more changes to support the current master for Boringssl from https://github.com/google/boringssl and created a new PR (https://github.com/lsh123/xmlsec/pull/240/)

**Pull request closed without merging**